harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            'please set `resources["optimized_parallel"] = True`.',
+        )
+
+        logging.info(f"Enabling mixed precision training with opt_level: {opt_level}.")
+        models, optimizers = apex.amp.initialize(
+            models=models,
+            optimizers=optimizers,
+            enabled=enabled,
+            opt_level=opt_level,
+            cast_model_type=cast_model_type,
+            patch_torch_functions=patch_torch_functions,
+            keep_batchnorm_fp32=keep_batchnorm_fp32,
+            master_weights=master_weights,
+            loss_scale=loss_scale,
+            cast_model_outputs=cast_model_outputs,
+            num_losses=num_losses,
+            min_loss_scale=min_loss_scale,
+            max_loss_scale=max_loss_scale,
+            verbosity=verbosity
+            if self.distributed.get_rank() == 0 or self.env.experiment_config.debug_enabled()
+            else 0,
+        )
+        if not isinstance(models, list):
+            self.models = [models]
+        if not isinstance(optimizers, list):
+            self.optimizers = [optimizers]
+        return models, optimizers
+
+    def _should_communicate_and_update(self) -> bool:
+        if self._current_batch_idx is None:
+            raise det.errors.InternalException("Training hasn't started.")
+        return (self._current_batch_idx + 1) % self.hvd_config.aggregation_frequency == 0
+
+    def backward(
+        self,
+        loss: torch.Tensor,
+        gradient: Optional[torch.Tensor] = None,
+        retain_graph: bool = False,
+        create_graph: bool = False,
+    ) -> None:
+        """Compute the gradient of current tensor w.r.t. graph leaves.
+
+        The arguments are used in the same way as ``torch.Tensor.backward``.
+        See https://pytorch.org/docs/1.4.0/_modules/torch/tensor.html#Tensor.backward for details.
+
+        Arguments:
+            gradient (Tensor or None): Gradient w.r.t. the
+                tensor. If it is a tensor, it will be automatically converted
+                to a Tensor that does not require grad unless ``create_graph`` is True.
+                None values can be specified for scalar Tensors or ones that
+                don't require grad. If a None value would be acceptable then
+                this argument is optional.
+            retain_graph (bool, optional): If ``False``, the graph used to compute
+                the grads will be freed. Note that in nearly all cases setting
+                this option to True is not needed and often can be worked around
+                in a much more efficient way. Defaults to the value of
+                ``create_graph``.
+            create_graph (bool, optional): If ``True``, graph of the derivative will
+                be constructed, allowing to compute higher order derivative
+                products. Defaults to ``False``.
+        """
+        if self._use_amp:
+            check.eq(
+                loss,
```

Comment:
question: won't this check always fail on the first call since `self._loss` is going to be `None`?